README.md
# node-storinfo

This repository is part of the Joyent Manta project.  For contribution
guidelines, issues, and general documentation, visit the main
[Manta](https://github.com/joyent/manta) project page.

node-storinfo is a NodeJS client for the [Manta Storinfo
service](https://github.com/joyent/manta-storinfo).

*NOTE* This module is load-bearing for both the muskie (webapi) and buckets-api
MantaV2 services.  Therefore, if you make changes to this node package, you
should sanity check the functionality of both services before integration.


## Creating a Storinfo Client

node-storinfo supports initializing either a standalone or full StorinfoClient
object.  A standalone client does not establish a network connection to the
Storinfo service.  It is primarily used for testing the object placement
algorithm implemented by the `choose` method without requiring access to an
actual Manta deployment.

A standalone client can be created as follows:

``` js
const mod_storinfo = require('storinfo');
           
var client = mod_storinfo.createClient({standalone: true});
```

A full client (one that actually connects to the Storinfo service in a Manta
deployment) requires a number of additional properties.

| property    | type   | description                                           |
| ----------- | ------ | ----------------------------------------------------- |
| url         | string | URL of Storinfo service                               |
| cueballOpts | object | Required parameters for creating a Cueball HttpAgent. |

For example:

``` js
var opts = {
    standalone: false,
    url: 'http://storinfo.domain'
    cueballOpts: {
        spares: 4,
        maximum: 10,
        recovery: {
            default: {
                timeout: 2000,
                retries: 5,
                delay: 250,
                maxDelay: 1000
            }
        },
        resolvers: ["nameservice.domain"]
    }
};

var client = mod_storinfo.createClient(opts);
```



Additionally, the following _optional_ parameters can be specified when
creating a StorinfoClient instance:

| property                  | type    | description                                                  |
| ------------------------- | ------- | ------------------------------------------------------------ |
| log                       | object  | Bunyan logger.  If specified, node-storinfo will write to this log and also pass it down to cueball and restify.  Otherwise, a new logger will be created. |
| multiDC                   | boolean | Specifies whether the Manta environment contains multiple data centers.  This affects the object placement algorithm implemented by the `choose` method.  In single DC test environments, this can set to false to override the requirement for spreading replicas across DCs.  Default is *false* |
| pollInterval              | number  | If specified, the storinfo client will invoke the GetStorageNodes API for the Storinfo service at the frequency specified by pollInterval.  The value is interpreted as milliseconds.  A 'topology' event will be emitted by the StorinfoClient object after each successful poll. |
| defaultMaxStreamingSizeMB | number  | The maximum allowed size (in MB) for a streaming upload. Default is 5120 MB                                           |
| maxUtilizationPct         | number  | The maximum storage node utilization threshold (as a percentage) for normal (non-operator) object writes.  This affects the object placement algorithm implemened by the `choose` method.  Default is 90.|



## StorinfoClient methods

#### getStorageNodes

The getStorageNodes method returns the Storinfo services cached view of the
entire `manta_storage` bucket, as an array of objects sorted by
`manta_storage_id`.  This method will return an error if invoked on a
standalone client.

This asynchronous method takes a the following argument:

| argument   | type     | description                                                  |
| ---------- | -------- | ------------------------------------------------------------ |
| callback(err, obj) | function | Callback to be invoked upon completion.  The callback will be invoked with two parameters: "res" and "err".  On success, the "err" parameter to the callback will be null and the "res" param will contain an array of for the requested storage node.  On failure, the "res" param will be null and "err" will contain a verror object describing the failure. |



#### getStorageNode

The getStorageNode method returns the Storinfo services cached view of a single
row from the `manta_storage` bucket, corresponding to the storage node with the
specified `manta_storage_id`.   This method will return an error if invoked on
a standalone client.

This asynchronous method takes a the following arguments:

| argument   | type     | description                                                  |
| ---------- | -------- | ------------------------------------------------------------ |
| storageid  | string   | Manta storage ID of the storage node on which to request storage utilization data |
| callback(err, obj) | function | Callback to be invoked upon completion.  The callback will be invoked with two parameters: "obj" and "err".  On success, the "err" parameter to the callback will be null and the "obj param will contain `manta_storage` object for the requested storage node.  On failure, the "obj" param will be null and "err' willc ontain a verror object describing the failure. |



#### choose

The choose method takes a desired number of replicas and a size (in bytes), and
then selects three random "tuples" (the number of items in a tuple is #replicas).
The first random tuple is "primary," and then we have 2 backup tuples.

Conceptually it looks like this:

```
{
    us-east-1: [a, b, c, ...],
    us-east-2: [d, e, f, ...],
    us-east-3: [g, h, i, ...],
    ...
}
```

Where the objects `a...N` are the full JSON representation of a single storage
node.

The choose method can be invoked for both standalone and for StorinfoClient
objects.  For non-standalone clients, this method requires that the
StorinfoClient has successfully performed at least on poll from the Storinfo
service.  This can be assured by either manually calling the `getStorageNodes`
method or by specifying the pollInterval property during client creation and
then waiting for a 'topology' event.

This method takes the argument object and a callback.  The properties of the
argument object are described below:

| argument   | type    | description                                                  |
| ---------- | ------- | ------------------------------------------------------------ |
| size       | number  | Size of the object (in MB) to be stored.   This is optional and defaults to 5120. |
| replicas   | number  | Number of copies of the object to store (i.e. x-durability-level).  This is optional and defaults to 2. |
| isOperator | boolean | Is this PUT request coming from an operator account?  This is optional and defaults to false. |

## mchoose CLI

This module includes a CLI `bin/mchoose` which provides a scriptable interface
for invoking the `getStorageNodes` and `choose` methods.  The usage is
described below:

```
Models the behavior of the Manta's object placement logic.

Usage:
    mchoose [OPTIONS] COMMAND [ARGS...]
    mchoose help COMMAND

Options:
    -h, --help      Show this help message and exit.
    --version       Print version and exit.

Commands:
    help (?)        Help on a specific sub-command.
    poll            Poll the Storinfo service for storage records and summarizes the results.
    choose          Simulate Manta storage node selection for an object.
```

## Testing

The automated tests do not require access to a Manta deployment.  They can be
run as follows:

```
% make test
```



CHANGES.md
# node-storinfo Changelog

## Not yet released.

None yet.

## 1.0.1

[MANTA-5134](https://smartos.org/bugview/MANTA-5134) Muskie SMF service failed to start up when it cannot find storinfo service

## 1.0.0

First release.


test/mchoose.test.js
/*
 * Copyright 2020 Joyent, Inc.
 */

/*
 * Test the "mchoose" command.
 */

const forkExecWait = require('forkexec').forkExecWait;
const path = require('path');

const test = require('tap').test;

const BINDIR = path.resolve(__dirname, '../bin');
const MCHOOSE = path.resolve(BINDIR, 'mchoose');


// ---- tests

/*
 * Verify command can be invoked without error
 */
test('mchoose -h', function (t) {
    var argv = [
        MCHOOSE,
        '-h'
    ];

    var usagePhrase =
        'Models the behavior of the Manta\'s object placement logic.';

    forkExecWait({
        argv: argv
    }, function (err, info) {
        t.ifError(err, err);

        t.equal(info.stderr, '', 'no stderr');
        t.equal(info.stdout.lastIndexOf(usagePhrase, 0), 0,
            'stdout from mchoose');

        t.end();
    });
});


test/storinfo.test.js
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

/*
 * Copyright 2020 Joyent, Inc.
 */

const assert = require('assert-plus');
const fs = require('fs');
const path = require('path');

const test = require('tap').test;

const mod_storinfo = require('../lib/client.js');

// /--- Constants

const THREE_SIGMA = 0.997;
const DEF_MAX_STREAMING_SIZE_MB = 5120;
const DEF_MAX_PERCENT_UTIL = 90;


// /--- Tests

/**
 * Sum values in array
 *
 * @param {!number[]} values
 */
function sum(values) {
    return values.reduce(function (acc, v) {
        acc += v;
        return (acc);
    }, 0);
}

/**
 * Calculate the arithmetic mean of an array of values
 *
 * @param {!number[]} values
 * @returns {number} the sum of values
 */
function mean(values) {
    return (sum(values) / values.length);
}

/**
 * Calculate the standard deviation of an array of values
 *
 * @param {!number[]} values
 * @param {number} [av] - mean, if already calculated
 * @returns {number} the standard deviation of values
 */
function sd(values, av) {
    if (av === undefined) {
        av = mean(values);
    }

    var sqd = values.map(function (v) {
        var res = v - av;
        return (res * res);
    });

    var meansqd = mean(sqd);
    return (Math.sqrt(meansqd));
}

/**
 * Calculate zscore of each element in array of values
 *
 * @param {!number[]} values
 * @param {number} [av] - mean, if already calculated
 * @param {number} [stdd] - standard deviation, if already calculated
 * @returns {number[]} an array of zscores for values
 */
function zscores(values, av, stdd) {
    if (av === undefined) {
        av = mean(values);
    }
    if (stdd === undefined) {
        stdd = sd(values, av);
    }
    return (values.map(function (v) {
        return ((v - av) / stdd);
    }));
}

/**
 * Rough test to see if values follow a normal distribution
 *
 * NOTE: this is non-deterministic, we could do with some Erlang quickcheck
 * ?SOMETIMES magic here (see http://quviq.com/documentation/eqc/ macro
 * ?SOMETIMES for details.)
 *
 * @param {object} kvs - an object of {string} -> {number} mappings
 * @returns {boolean} true if the numbers are most likely normally
 * distributed
 */
function probablyNormal(kvs) {
    var values = Object.keys(kvs).map(function (k) {
        return (kvs[k]);
    });
    var zs = zscores(values);

    // outliers are greater than 3sd from the mean
    var outliers = zs.filter(function (z) {
        return (Math.abs(z) > 3);
    });

    // NOTE: in this case, probably outliers.length === 0 would be
    // enough
    return ((outliers.length / values.length) < (1 - THREE_SIGMA));
}

test('storinfo records', function (t) {
    // report vars (see report below)
    var min = Infinity;
    var max = 0;
    var total = 0;
    var runs = 0;
    var dcs = {};
    var hosts = {};

    /*
     * Test iterations. Pulled from old lib/picker.js test from the
     * manta-muskie repotest, see blame/history on that file.
     */
    var N = 10000;

    // read sub/mock/fake moray results
    var fname = 'storinfo.records.json';
    var file = path.join(__dirname, '..', 'test', fname);
    var morayValuesJSON = fs.readFileSync(file, 'utf8');
    var morayValues = JSON.parse(morayValuesJSON);

    /*
     * As we're using the canned storage node data in storinfo.records.json,
     * we create a standalone storinfo client.
     */
    var storinfo = mod_storinfo.createClient({
        log: require('bunyan').createLogger({
            level: process.env.LOG_LEVEL || 'info',
            name: 'storinfo_test',
            stream: process.stdout
        }),
        defaultMaxStreamingSizeMB: DEF_MAX_STREAMING_SIZE_MB,
        maxUtilizationPct: DEF_MAX_PERCENT_UTIL,
        multiDC: true,
        standalone: true
    });

    // set up the storinfo with the fake moray data
    mod_storinfo.sortAndStoreDcs.call(storinfo, morayValues, morayValues);

    // do it
    select(t);

    // NOTE: only reports on failure, to aid debugging
    function report() {
        var hostsNormal = probablyNormal(hosts);
        var dcsNormal = probablyNormal(dcs);

        if (!(hostsNormal && dcsNormal)) {
            var keys = Object.keys(hosts);
            keys.sort();
            console.log('**** Host Selection ****');
            keys.forEach(function (k) {
                console.log(k + ' ' + hosts[k]);
            });
            console.log('\n**** Datacenter Selection ****');
            keys = Object.keys(dcs);
            keys.sort();
            keys.forEach(function (k) {
                console.log(k + ' ' + dcs[k]);
            });
            console.log('\n**** Timing ****');
            console.log('avg: ' + total / N  + 'ms');
            console.log('max: ' + max + 'ms');
            console.log('min: ' + min + 'ms');
        }

        storinfo.close();
        t.ok(hostsNormal, 'Hosts selection appears normally distributed');
        t.ok(dcsNormal, 'DC selection appears normally distributed');
        t.end();
    }

    // the actual test code, yes, it runs iterations
    function select() {
        var start = new Date().getTime();
        storinfo.choose({}, function onChosen(err, sharks) {
            assert.ifError(err);
            var delta = new Date().getTime() - start;
            if (delta > max) {
                max = delta;
            }
            if (delta < min) {
                min = delta;
            }
            total += delta;

            Object.keys(sharks).forEach(function track(k) {
                var dc_names = [];
                sharks[k].forEach(function (s) {
                    var id = s.manta_storage_id;
                    if (!hosts[id]) {
                        hosts[id] = 0;
                    }

                    hosts[id]++;

                    dc_names.push(s.datacenter);
                });

                var dc_key = dc_names.join(' ');
                if (!dcs[dc_key]) {
                    dcs[dc_key] = 0;
                }
                dcs[dc_key]++;
            });

            setImmediate((++runs < N ? select : report), t);
        });
    }

});


lib/errors.js
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

/*
 * Copyright 2020 Joyent, Inc.
 */

module.exports = {
    NotEnoughSpaceError: NotEnoughSpaceError,
    StandaloneModeError: StandaloneModeError
};

const mod_util = require('util');
const mod_verror = require('verror');
const VError = mod_verror.VError;

function NotEnoughSpaceError(size, cause) {
    var opts = {};
    opts.constructorOpt = NotEnoughSpaceError;
    this.size = size;
    this.cause = cause;
    VError.call(this, opts, 'Unable to place object (size: %s MB, cause: %s)',
        size, cause);
}
mod_util.inherits(NotEnoughSpaceError, VError);
NotEnoughSpaceError.prototype.name = 'NotEnoughSpaceError';

function StandaloneModeError() {
    var opts = {};
    opts.constructorOpt = StandaloneModeError;
    VError.call(this, opts, 'this method is not supported in standalone mode');
}
mod_util.inherits(StandaloneModeError, VError);
StandaloneModeError.prototype.name = 'StandaloneModeError';


lib/client.js
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

/*
 * Copyright 2020 Joyent, Inc.
 */

const EventEmitter = require('events').EventEmitter;

const assert = require('assert-plus');
const bunyan = require('bunyan');
const cueball = require('cueball');
const errors = require('./errors.js');
const once = require('once');
const restify = require('restify-clients');
const util = require('util');

const sprintf = util.format;

const DEF_NUM_COPIES = 2;
const DEF_MAX_STREAMING_SIZE_MB = 5120;
const DEF_MAX_PERCENT_UTIL = 90;

/*
 * Fisher-Yates shuffle - courtesy of http://bost.ocks.org/mike/shuffle/
 *
 * Called by StorinfoClient.choose()
 */
function shuffle(array) {
    var m = array.length,
        t, i;

    while (m) {
        i = Math.floor(Math.random() * m--);
        t = array[m];
        array[m] = array[i];
        array[i] = t;
    }
    return (array);
}

/*
 * Just picks a random number, and optionally skips the last one we saw.
 *
 * Called by StorinfoClient.choose()
 */
function random(min, max, skip) {
    var num = (Math.floor(Math.random() * (max - min + 1)) + min);

    if (num === skip) {
        num = ((num + 1) % max);
    }

    return (num);
}

/*
 * Modified binary-search. We're looking for the point in the set at which all
 * servers have at least the requested amount of space.  Logically you would
 * then do set.slice(lower_bound(set, 100));
 * But that creates a copy - but really the return value of this to $end is
 * what the choose logic can then look at.
 *
 * Called by StorinfoClient.choose()
 */
function lower_bound(set, size, low, high) {
    assert.arrayOfObject(set, 'set');
    assert.number(size, 'size');
    assert.optionalNumber(low, 'low');

    low = low || 0;
    high = high || set.length;

    while (low < high) {
        var mid = Math.floor(low + (high - low) / 2);
        if (set[mid].availableMB >= size) {
            high = mid;
        } else {
            low = mid + 1;
        }
    }

    if (!set[low] || set[low].availableMB < size) {
        low = -1;
    }

    return (low);
}

/*
 * A comparison function used to order storage zones based on available space.
 *
 * @param {object} a               - a storage zone object
 * @param {integer} a.availableMB  - free space in MB on the storage zone
 * @param {object} b               - a storage zone object
 * @param {integer} b.availableMB  - free space in MB on the storage zone
 * @throws {TypeError} on bad input.
 */
function storageZoneComparator(a, b) {
    assert.object(a, 'a');
    assert.object(b, 'b');
    assert.number(a.availableMB, 'a.availableMB');
    assert.number(b.availableMB, 'b.availableMB');

    if (a.availableMB < b.availableMB) {
        return (-1);
    } else if (a.availableMB > b.availableMB) {
        return (1);
    }

    return (0);
}

/*
 * A function to sort the storage zones available for normal requests and those
 * available only for operator requests within each datacenter by available
 * storage.
 *
 * @param {object} dcObj   - an object mapping datacenters to their associated
 *                           storage zones
 * @param {object} opDcObj - an object mapping datacenters to their associated
 *                           storage zones
 * @throws {TypeError} on bad input.
 */
function sortAndStoreDcs(dcObj, opDcObj) {
    assert.object(dcObj, 'dcObj');
    assert.object(opDcObj, 'opDcObj');

    var dcCount = 0;
    var operatorDcCount = 0;
    var dcs = Object.keys(dcObj);
    var operatorDcs = Object.keys(opDcObj);

    dcs.forEach(function dcSortAndCount(k) {
        dcObj[k].sort(storageZoneComparator);
        dcCount++;
    });

    operatorDcs.forEach(function opDcSortAndCount(k) {
        opDcObj[k].sort(storageZoneComparator);
        operatorDcCount++;
    });

    if (dcCount > 0) {
        this.datacenters = dcs;
    } else {
        this.log.warn('sortAndStoreDcs: could not find any minnow ' +
            'instances');
        this.datacenters = [];
    }

    if (operatorDcCount > 0) {
        this.operatorDatacenters = operatorDcs;
    } else {
        this.log.warn('sortAndStoreDcs: could not find any minnow ' +
            'instances for operator requests');
        this.operatorDatacenters = [];
    }

    this.dcSharkMap = dcObj;
    this.operatorDcSharkMap = opDcObj;
    this.emit('topology', [this.dcSharkMap, this.operatorDcSharkMap]);

    this.log.trace('sortAndStoreDcs: done');
}

function doPoll() {
    clearTimeout(this.pollTimer);

    var self = this;
    var args = {};

    self.getStorageNodes(args, function (err, res) {
        this.pollTimer = setTimeout(doPoll.bind(self), self.pollInterval);
        if (err) {
            self.log.error(err, 'doPoll: unexpected error ' +
                '(will retry)');
            return;
        }

        var dcObj = {};
        var opDcObj = {};

        function sortByDatacenter(maxUtilization, v) {
            if (!opDcObj[v.datacenter]) {
                opDcObj[v.datacenter] = [];
            }

            /*
             * The Storinfo service's /storagenodes interface already filters
             * out storage nodes who's utilization is above the operator
             * threshold.
             */
            opDcObj[v.datacenter].push(v);

            /*
             * Moray is queried for the sharks whose utilization is less than
             * or equal to the maximum utilization percentage at which operator
             * writes are still accepted. Find the set of sharks whose
             * utilization is less than or equal to the utilization threshold
             * for all requests.
             */
            if (v.percentUsed <= maxUtilization) {
                if (!dcObj[v.datacenter]) {
                    dcObj[v.datacenter] = [];
                }

                dcObj[v.datacenter].push(v);
            }
        }

        res.forEach(sortByDatacenter.bind(self, self.utilization));

        /*
         * We just defer to the next tick so we're not tying
         * up the event loop to sort a lot if the list is large
         */
        setImmediate(sortAndStoreDcs.bind(self, dcObj, opDcObj));
    });
}

/*
 * url: storinfo service URL
 * cueballOpts: object containing required cueball options
 */
function StorinfoClient(opts) {
    assert.object(opts, 'opts');
    assert.optionalObject(opts.log, 'log');
    assert.optionalNumber(opts.pollInterval, 'pollInterval');
    assert.optionalBool(opts.multiDC, 'multiDC');
    assert.optionalNumber(opts.defaultMaxStreamingSizeMB,
        'defaultMaxStreamingSizeMB');
    assert.optionalNumber(opts.maxUtilizationPct, 'maxUtilizationPct');
    assert.bool(opts.standalone, 'standalone');

    this.log = opts.log || bunyan.createLogger({ name: 'storinfo' });

    /*
     * If we're not running in standalone mode, create the restify client.
     */
    if (!opts.standalone) {
        assert.string(opts.url, 'url');
        assert.object(opts.cueballOpts, 'cueballOpts');

        opts.cueballOpts.log = this.log;
        var clientOpts = {
            url: opts.url,
            agent: new cueball.HttpAgent(opts.cueballOpts),
            log: this.log
        };

        this.http = restify.createJsonClient(clientOpts);
    }
    this.standalone = opts.standalone;

    EventEmitter.call(this);

    /*
     * The dcSharkMap is an object that maps datacenter names to an array of
     * sharks sorted by available storage capacity that are all at or below the
     * storage utilization threshold for normal manta requests.
     */
    this.dcSharkMap = null;
    /*
     * The operatorDcSharkMap is an object that maps datacenter names to an
     * array of sharks sorted by available storage capacity that are all at or
     * below the storage utilization threshold for operator manta requests.
     */
    this.operatorDcSharkMap = null;

    this.dcIndex = -1;
    this.multiDC = opts.multiDC === undefined ? true : opts.multiDC;
    this.datacenters = null;
    this.defMaxSizeMB = opts.defaultMaxStreamingSizeMB ||
        DEF_MAX_STREAMING_SIZE_MB;
    this.utilization = opts.maxUtilizationPct ||
        DEF_MAX_PERCENT_UTIL;

    if (opts.pollInterval !== undefined && !opts.standalone) {
        this.pollInterval = opts.pollInterval;
        setImmediate(doPoll.bind(this));
    }
}
util.inherits(StorinfoClient, EventEmitter);

StorinfoClient.prototype.close = function close() {
    if (this.pollTimer) {
        clearTimeout(this.pollTimer);
    }
    if (this.http) {
        this.http.close();
    }
};

/*
 * Get the data for a specific row (storage node) from the manta_storage bucket.
 * There are two required arguments.  The first is the manta_storage_id of the
 * storage node.  The second is a callback withthew signature:
 *   callback(err, obj)
 *
 * On success, the "err" parameter to the callback will be null and the "obj"
 * param will contain manta_storage object for the requested storage node.
 *
 * On failure, the "obj" param will be set to null.
 */
StorinfoClient.prototype.getStorageNode = function getStorageNode(storageid,
    callback) {

    assert.string(storageid, 'storageid');

    /*
     * getStorageNode isn't supported in standalone mode
     */
    if (this.standalone) {
        var standErr = new errors.StandaloneModeError();
        callback(standErr, null);
        return;
    }

    this.http.get('/storagenodes/' + storageid, function (err, req, res, obj) {
        if (err) {
            callback(err, null);
            return;
        }
        callback(null, obj);
    });
};

/*
 * This function wraps the /storagenodes API provided by the Manta Storinfo
 * service.
 * It takes two arguemnts:
 *   args: object containing the following properties
 *       opts: optional object containing the following optional params:
 *           after_id: String manta_storage_id
 *           limit: Number
 *
 *       The args object is also used internally to store the intermiediate and
 *       final results of successive calls to /storagenodes.
 *
 *   callback: callback with the signature: callback(err, res)
 *       err: object containing error details if an error occurs
 *            on success, this param will be null
 *       res: on success, this will contain the response from /storagenodes
 *            on error, this will be null
 */
StorinfoClient.prototype.getStorageNodes = function getStorageNodes(args,
    callback) {

    assert.object(args, 'args');
    assert.optionalObject(args.opts);
    if (args.opts) {
        assert.optionalString(args.opts.after_id, 'opts.after_id');
        assert.optionalFinite(args.opts.limit, 'opts.limit');
    }

    var self = this;

    /*
     * getStorageNodes isn't supported in standalone mode
     */
    if (self.standalone) {
        var standErr = new errors.StandaloneModeError();
        callback(standErr, null);
        return;
    }

    var httpOpts = {
        path: '/storagenodes',
        query: args.opts
    };

    if (args.res === undefined) {
        args.res = [];
    }

    this.http.get(httpOpts, function (err, req, res, obj) {
        if (err) {
            callback(err, null);
            return;
        }
        Array.prototype.push.apply(args.res, obj);
        if (res.headers.link !== undefined) {
            /*
             * The format of the 'link' is:
             * </storagenodes?after_id=storageid&limiit=number>; rel=\"next\"
             *
             * we extract the after_id parameter
             */
            args.opts.after_id =
                res.headers.link.split('=')[1].split('&')[0];
            self.getStorageNodes(args, function (_r, e) {
                callback(e, args.res);
            });
        } else {
            callback(err, args.res);
        }
    });
};


/*
 * Selects N shark nodes from sharks with more space than the request length.
 *
 * @param {object} options -
 *                   - {number} size => req.getContentLength()
 *                   - {number} replicas => req.header('x-durability-level')
 *                   - {boolean} isOperator => req.caller.account.isOperator
 * @param {funtion} callback => f(err, [sharkClient])
 *
 * Choose takes a desired number of replicas and a size (in bytes), and then
 * selects three random "tuples" (the number of items in a tuple is
 * #replicas).  The first random tuple is "primary," and then we have 2 backup
 * tuples.  The contract here is that upstack code tries all hosts in
 * "primary," and if all are up we're good to go; if any fail it falls through
 * to trying all hosts in "secondary." While not the most sophisticated and/or
 * error-proof approach, this is simple to reason about, and should be "good
 * enough," given what we know about our infrastructure (i.e., we expect it to
 * be up).
 *
 * So in terms of implementation, Storinfo periodically refreshes a (sorted) set
 * of servers per datacenter that is advertised in a moray bucket
 * (manta_storage).  To see how data gets in manta_storage, see minnow.git.
 *
 * So conceptually it looks like this:
 *
 * {
 *   us-east-1: [a, b, c, ...],
 *   us-east-2: [d, e, f, ...],
 *   us-east-3: [g, h, i, ...],
 *   ...
 * }
 *
 * Where the objects `a...N` are the full JSON representation of a single mako
 * instance.  In that object, we really only care about two fields:
 *
 *   -- manta_storage_id (hostname)
 *   -- availableMB
 *
 * We keep those sets sorted by `availableMB`, and everytime choose is run, we
 * make a "view" of the set for each data center that tells us all the servers
 * that have that amount of storage and larger (binary search).
 *
 * Once we have that "view," we simply pick random nodes from the set(s).
 * Lastly, we RR across DCs so we spread objects around evenly.
 */
StorinfoClient.prototype.choose = function choose(opts, cb) {
    assert.object(opts, 'options');
    assert.optionalObject(opts.log, 'options.log');
    assert.optionalNumber(opts.replicas, 'options.replicas');
    assert.optionalNumber(opts.size, 'options.size');
    assert.optionalBool(opts.isOperator, 'options.isOperator');
    assert.func(cb, 'callback');

    cb = once(cb);

    var dcs = [];
    var log = this.log;
    var offsets = [];
    var replicas = opts.replicas || DEF_NUM_COPIES;
    var seen = [];
    var self = this;
    var size = Math.ceil((opts.size || 0) / 1048576) || this.defMaxSizeMB;
    var err, cause;

    log.debug({
        replicas: replicas,
        size: size,
        defMaxSizeMB: this.defMaxSizeMB
    }, 'StorinfoClient.choose: entered');

    /*
     * Determine the index of the first storage node for each DC that has space
     * for an object of the requested size.  If no sharks in a given DC have
     * enough space, we exclude them from the possible set of DCs to choose
     * from.
     */
    function filterDatacenters(sharkMap, dc) {
        var l = lower_bound(sharkMap[dc], size);
        if (l !== -1) {
            dcs.push(dc);
            offsets.push(l);
        }
    }

    var filterFun;

    if (opts.isOperator) {
        filterFun = filterDatacenters.bind(this, this.operatorDcSharkMap);
        this.operatorDatacenters.forEach(filterFun);
    } else {
        filterFun = filterDatacenters.bind(this, this.dcSharkMap);
        this.datacenters.forEach(filterFun);
    }

    var chooseStats = {
        db: opts.isOperator ? self.operatorDcSharkMap : self.dcSharkMap,
        dcsInUse: dcs,
        offsets: offsets
    };

    var enoughDCs = false;
    if (dcs.length === 0) {
        cause = sprintf('no DC with sufficient space');
    } else if (replicas > 1 && this.multiDC && dcs.length < 2) {
        cause = sprintf('%d copies requested, but only %d DC(s) have ' +
            'sufficient space', replicas, dcs.length);
    } else {
        enoughDCs = true;
    }

    if (!enoughDCs) {
        log.warn('StorinfoClient.choose: not enough DCs available');
        assert.string(cause, 'no error message set');

        err = new errors.NotEnoughSpaceError(size, cause);
        cb(err, null, chooseStats);
        return;
    }

    dcs = shuffle(dcs);

    /*
     * Pick a random shark from the next DC in the round robin ordering.  If it
     * hasn't yet been used for a set, return the shark.
     *
     * If the shark has been chosen for another set, iterate through all sharks
     * in the DC until we find one that hasn't yet been seen.
     *
     * If there are no sharks that haven't yet been used in the DC, return null.
     */
    function host() {
        if (++self.dcIndex >= dcs.length) {
            self.dcIndex = 0;
        }

        var ndx = self.dcIndex;
        var dc;
        if (opts.isOperator) {
            dc = self.operatorDcSharkMap[dcs[ndx]];
        } else {
            dc = self.dcSharkMap[dcs[ndx]];
        }

        var s = random(offsets[ndx], dc.length - 1);

        if (seen.indexOf(dc[s].manta_storage_id) === -1) {
            seen.push(dc[s].manta_storage_id);
        } else {
            var start = s;
            do {
                if (++s === dc.length) {
                    s = offsets[ndx];
                }

                if (s === start) {
                    log.debug({
                        datacenter: dcs[ndx]
                    }, 'StorinfoClient.choose: exhausted DC');
                    return (null);
                }

            } while (seen.indexOf(dc[s].manta_storage_id) !== -1);

            seen.push(dc[s].manta_storage_id);
        }

        return ({
            datacenter: dc[s].datacenter,
            manta_storage_id: dc[s].manta_storage_id
        });
    }

    /*
     * Return a set with `replicas` sharks.
     */
    function set() {
        var s = [];

        for (var j = 0; j < replicas; j++) {
            var _s = host();
            if (_s === null) {
                return (null);
            }
            s.push(_s);
        }

        return (s);
    }

    function mapFun(s) {
        return (s.datacenter);
    }

    function reduceFun(last, now) {
        if (last.indexOf(now) === -1) {
            last.push(now);
        }
        return (last);
    }

    /*
     * We always pick three sets, and we pedantically ensure that we've got
     * them splayed x-dc
     */
    var sharks = [];
    for (var i = 0; i < 3; i++) {
        var tuple = set();

        if (!sharks.length && (!tuple || tuple.length < replicas)) {
            cause = 'copies requested exceeds number of available ' +
                'storage nodes';
            err = new errors.NotEnoughSpaceError(size, cause);
            cb(err, null, chooseStats);
            return;
        } else if (tuple && this.multiDC && replicas > 1) {
            var _dcs = tuple.map(mapFun).reduce(reduceFun, []);

            if (_dcs.length < 2) {
                cause = 'insufficient number of DCs selected';
                err = new errors.NotEnoughSpaceError(size, cause);
                cb(err, null, chooseStats);
                return;
            }
        }

        if (tuple) {
            sharks.push(tuple);
        }
    }

    log.debug({
        replicas: replicas,
        sharks: sharks,
        size: size
    }, 'StorinfoClient.choose: done');
    cb(null, sharks, chooseStats);
};


module.exports = {
    StorinfoClient: StorinfoClient,
    createClient: function (opts) {
        return (new StorinfoClient(opts));
    },
    sortAndStoreDcs: sortAndStoreDcs
};
